cli/command.go

```diff
@@ -0,0 +1,282 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/mattn/go-isatty"
+)
+
+// Command is the interface for a command or subcommand. Most of these functions
+// have default implementations on [BaseCommand].
+type Command interface {
+	// Desc provides a short, one-line description of the command. It should be
+	// shorter than 50 characters.
+	Desc() string
+
+	// Help is the long-form help output. It should include usage instructions and
+	// flag information.
+	Help() string
+
+	// Hidden indicates whether the command is hidden from help output.
+	Hidden() bool
+
+	// Run executes the command.
+	Run(ctx context.Context, args []string) error
+
+	Prompt(msg string) (string, error)
+
+	// Stdout returns the stdout stream. SetStdout sets the stdout stream.
+	Stdout() io.Writer
+	SetStdout(w io.Writer)
+
+	// Stderr returns the stderr stream. SetStderr sets the stderr stream.
+	Stderr() io.Writer
+	SetStderr(w io.Writer)
+
+	// Stdin returns the stdin stream. SetStdin sets the stdin stream.
+	Stdin() io.Reader
+	SetStdin(r io.Reader)
+
+	// Pipe creates new unqiue stdin, stdout, and stderr buffers, sets them on the
+	// command, and returns them. This is most useful for testing where callers
+	// want to simulate inputs or assert certain command outputs.
+	Pipe() (stdin, stdout, stderr *bytes.Buffer)
+}
+
+// CommandFactory returns a new instance of a command. This returns a function
+// instead of allocations because we want the CLI to load as fast as possible,
+// so we lazy load as much as possible.
+type CommandFactory func() Command
+
+// Ensure [RootCommand] implements [Command].
+var _ Command = (*RootCommand)(nil)
+
+// RootCommand represents a command root for a parent or collection of
+// subcommands.
+type RootCommand struct {
```

Comment:
Not super important, but this structure probably won't allow "persisting" flags easily, will it? Like if I have a set of flags commonly shared for all sub commands.

Reply:
Where should I define the flag "targets" (those `flagXxx` in code)? In JVS example, I see them defined in the corresponding command struct. If I want to share a set of flags, should I create a separate struct to hold those flags? Or define them as global vars?